src/Hl7.Fhir.Core/Serialization/SourceHelpers.cs

```diff
@@ -16,52 +15,334 @@ public static string Truncate(string str)
+                // -HH:mm
+
+                if (firstCharacter != '+' && firstCharacter != '-')
+                {
+                    return false;
+                }
+                var negative = firstCharacter == '-';
+                _position++;
+                if (!TryGetNumber(2, out hour, out var _)
+                    || hour > 14)
+                {
+                    return false;
+                }
+                if (!TrySkipCharacter(':'))
+                {
+                    return false;
+                }
+                if (!TryGetFixedNumber(2, out minute)
+                    || minute > 59
+                    || hour >= 14 && minute > 0)
+                {
+                    return false;
+                }
+                if (negative)
+                {
+                    hour = -hour;
+                    minute = -minute;
+                }
+                return true;
+            }
+
+            private bool AtEnd()
+            {
+                return _position >= _dateTimeString.Length;
+            }
+
+            private bool TrySkipCharacter(char character)
+            {
+                if (_position >= _dateTimeString.Length || _dateTimeString[_position] != character)
+                {
+                    return false;
+                }
+                _position++;
+                return true;
+            }
+
+            private bool TryGetYear(out int year)
+            {
+                return TryGetFixedNumber(4, out year)
+                    && 1 <= year;
+            }
+
+            private bool TryGetMonth(out int month)
+            {
+                return TryGetFixedNumber(2, out month)
+                    && 1 <= month && month <= 12;
+            }
+
+            private bool TryGetDay(int year, int month, out int day)
+            {
+                return TryGetFixedNumber(2, out day)
+                    && 1 <= day && day <= GetMonthDays(year, month);
+            }
+
+            private bool TryGetFixedNumber(int numberOfDigits, out int value)
+            {
+                return TryGetNumber(numberOfDigits, out value, out var actualNumberOfDigits)
+                    && numberOfDigits == actualNumberOfDigits;
+            }
+
+            private bool TryGetNumber(int maxNumberOfDigits, out int value, out int numberOfDigits)
+            {
+                value = 0;
+                var startPosition = _position;
+                var maxPosition = Math.Min(_dateTimeString.Length, startPosition + maxNumberOfDigits);
+                while (_position < maxPosition && char.IsDigit(_dateTimeString[_position]))
+                {
+                    value = value * 10 + (_dateTimeString[_position] - '0');
+                    _position++;
+                }
+                numberOfDigits = _position - startPosition;
+                return _position > startPosition;
+            }
+
+            private static int GetMonthDays(int year, int month)
```

Comment:
any reason to not use the built in method [DateTime.DaysInMonth](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.daysinmonth?view=netframework-4.8.1)

Reply:
None beside ignorance